src/Hazel/Var.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- | An adaptation of the locally nameless representation (See Charguéraud's
+-- "The Locally Nameless Representation" for more information) which uses 2D
+-- de Bruijn variables.
+
+module Hazel.Var where
+
+import Control.Lens (Iso', Traversal', iso)
+import Data.Word (Word32)
+
+import qualified Data.Text as T
+
+-- | The distance in binding levels between a variable and its binder. The
+-- "first dimension" in our 2D de Bruijn scheme.
+newtype Depth
+  = Depth { _depth :: Word32 }
+  deriving (Eq, Show, Enum)
+
+unDepth :: Iso' Depth Word32
+unDepth = iso (\ (Depth d) -> d) Depth
+{-# INLINE unDepth #-}
+
+-- | A binder slot. The second dimension in our 2D de Bruijn scheme.
+newtype Slot
+  = Slot { _slot :: Word32 }
+  deriving (Eq, Show)
+
+unSlot :: Iso' Slot Word32
+unSlot = iso (\ (Slot s) -> s) Slot
```

Comment:
Also for all of these it would be possible to write shorter things like `iso _slot Slot`, which might be clearer.